pdr_backend/data_eng/gql_data_factory.py

```diff
@@ -0,0 +1,267 @@
+import os
+from typing import Dict, Callable
+
+from enforce_typing import enforce_types
+import polars as pl
+
+from pdr_backend.ppss.data_pp import DataPP
+from pdr_backend.ppss.data_ss import DataSS
+from pdr_backend.ppss.web3_pp import Web3PP
+from pdr_backend.util.timeutil import pretty_timestr, current_ut
+
+from pdr_backend.data_eng.plutil import (
+    has_data,
+    newest_ut,
+)
+
+from pdr_backend.util.subgraph_predictions import (
+    get_all_contract_ids_by_owner,
+)
+
+from pdr_backend.data_eng.gql.predictoor.predictions import (
+    predictions_schema,
+    get_predictoor_predictions_df,
+)
+
+
+@enforce_types
+class GQLDataFactory:
+    """
+    Roles:
+    - From each GQL API, fill >=1 parquet_dfs -> parquet files data lake
+    - From parquet_dfs, calculate stats and other dfs
+    - All timestamps, after fetching, are transformed into milliseconds wherever appropriate
+
+    Finally:
+       - "timestamp" values are ut: int is unix time, UTC, in ms (not s)
+       - "datetime" values ares python datetime.datetime, UTC
+    """
+
+    def __init__(self, pp: DataPP, ss: DataSS, web3: Web3PP):
+        self.pp = pp
+        self.ss = ss
+        self.web3 = web3
+
+        # TO DO: Solve duplicates from subgraph.
+        # Method 1: Cull anything returned outside st_ut, fin_ut
+        self.debug_duplicate = False
+
+        # TO DO: This code has DRY problems. Reduce.
+        # get network
+        if "main" in self.web3.network:
+            network = "mainnet"
+        elif "test" in self.web3.network:
+            network = "testnet"
+        else:
+            raise ValueError(self.web3.network)
+
+        # filter by feed contract address
+        contract_list = get_all_contract_ids_by_owner(
+            owner_address=self.web3.owner_addrs,
+            network=network,
+        )
+        contract_list = [f.lower() for f in contract_list]
+
+        # TO-DO: Roll into yaml config
```

Comment:
```Let's not add technical debt in each PR.```

I would rather restructure the whole code base into something that is functional, before weaving it into the yaml structure and then having to re-write that 3 times.

What i'm trying to avoid is optimizing early....

Reply:
Any new feature that needs settings needs yaml out of the gate. Not something to defer, because it's technical debt.